feat(Cards): Adding Cards components

gilad215 opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->
**What**: fix #192 #206

Creating the Cards components. Started with creating the mutual Base Card components:
`Card` `CardContainer` `CardHeading` `CardTitle` `CardBody` `CardLink` `CardDropdownButton` `CardFooter`

Continued with Aggregate Status Card:
`AggregateStatusCount` `AggregateStatusNotifications` `AggregateStatusNotification`
<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
**Link to Storybook**:
https://rawgit.com/gilad215/patternfly-react/storybook/feature/cards/index.html

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:
TODO:

- [x] #207

- [x] #208 

<!-- feel free to add additional comments -->

gilad215 commented:
Added **Trend Card** , **Utilization Trend Card**.

Writing the storybook page for the **Trend Card** didn't require any additional components.
I need clarification about #208 , @serenamarie125 linked to the **Utilization Trend Card** but named the issue after **Utilization Bar Card**, as for the Bard Card, I saw there is no react implementation for the progress bars (those that have tooltips) so i need more information on how to approach this.

Thanks

priley86 commented:
great job pulling this together so fast @gilad215!

I am seeing a few design issues with the knobs on the Basic Card in your Storybook. Can you look at the "accent" and "aggregate" knobs if you are intending to resolve those issues with this PR?

priley86 commented:
@gilad215 i would rather not conflate #207 and #208 in this PR yet... but it looks like you have done a nice job starting on those. I think #208 will have to wait for Utilization Bar PR in progress...

For Base Card, I think you can remove those aggregate knobs. Can you also show an additional card having this design variation:
![screen shot 2018-02-06 at 10 15 27 am](https://user-images.githubusercontent.com/4237045/35867372-94899f12-0b27-11e8-80f1-20306818c0e9.png)
Just want to make sure it works OK...

It would also be nice to demonstrate the cards in a `row-cards-pf` row and tackle the equal sizing with React Equalizer (so maybe you can render those other variations I mention). Maybe you can demonstrate that in another story called "Card Rows w/ Equalizer" ? I started on that in mIQ here and I think it is the simplest solution for now: 
https://github.com/priley86/miq_v2v_ui_plugin/blob/master/app/javascript/migration/scenes/Overview/index.js#L81

jeff-phillips-18 commented:
The `href={'#'}` on the links in the stories mess up the storybook navigation.

jeff-phillips-18 commented:
For tooltips in the stories we should use the Tooltip component and remove the 
```
                    data-toggle="tooltip"
                    data-placement="top"
```
references

gilad215 commented:
- Added "View CPU Events" Card to the Base Card Story.
- Fixed Aggregate Card Links @serenamarie125  
- Added Tooltips components and fixed links so the user stays in navigation @jeff-phillips-18

As for the `react-equalizer`, i added a story page for it but it doesn't seem to work. looks like the CardBody doesn't have the style child that the equalizer is looking for even though i'm giving it Data:
![undefined](https://user-images.githubusercontent.com/24938324/36075696-30747938-0f5b-11e8-8217-39f53500dd72.png)

I tried changing the CardBody to a regular div, equalizer got the style child it looked for but it still didn't change the heights as it should.
I uploaded the changes anyway to the BaseCardEqualizerStory so maybe we can find the issue together!

priley86 commented:
hey @gilad215 ... taking a look at this now ;)

priley86 commented:
couple things here:

* can you please use `innerRefs` for react equalizer (I think it is a bit cleaner)... guide here:
https://reactjs.org/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components
* let's try to make Equalizer optional (by also exposing `equalizer` boolean prop on the  child nodes)... some users may prefer to use pure css flexbox to do this in the future...
* it seems that some of the html structure is not conventional to me, but will have to check with @jeff-phillips-18 and @jgiardino 

will add these comments inline now so it is easier to follow...

priley86 commented on src/components/Cards/CardContainer.js:
```diff
@@ -0,0 +1,23 @@
+const CardContainer = ({ nodes, children, className, ...props }) => {
+  const classes = ClassNames('container-fluid container-cards-pf', className);
+
+  return (
```
i think we can expose a boolean `equalizer` prop and then render something like:
```
  if(equalizer){
  return (
    <Equalizer nodes={nodes} className={classes} {...props}>
      {children}
    </Equalizer>
  ); } else {
    <div className={classes} {...props}>
      {children}
    </div>
}
```

priley86 commented on src/components/Cards/Stories/BaseCardEqualizerStory.js:
```diff
@@ -0,0 +1,158 @@
+      flex: 1
+    };
+    return (
+      <Row
```
i need some designer help here @jeff-phillips-18 @jgiardino ... should `CardContainer` be wrapped within `Row`? I thought it was typically on the outside...
```
      <Row
        className={'row-cards-pf'}
        style={{ overflow: 'auto', paddingBottom: 100, height: '100%' }}
      >
        <CardContainer nodes={getNodes}>
          <Col xs={6} sm={4} md={4}>
```

gilad215 replied:
I did this because i wanted to follow your Equalizer [example](https://github.com/priley86/miq_v2v_ui_plugin/blob/master/app/javascript/migration/scenes/Overview/index.js#L81) exactly as you wrote it so i won't miss anything, i ran into a lot of problems with this package.

priley86 replied:
Yup .. my example was vaporware...It would be good to formalize it now though ;)

priley86 commented:
@serenamarie125 @jeff-phillips-18 @jgiardino i have not had a chance to review the Empty State Card here - can someone review the design of this?

gilad215 commented:
- Switched to `innerRefs` and now Equalizer works with the Card components :)
- Reverted back to functional components - `Card` `CardBody` `BaseCardEqualizerStory`.
`CardContainer` `Card` `CardBody` now accept `equalized` arg as seen in the storybook, in order to render the Cards properly (Using Flex CSS)

priley86 commented:
@jeff-phillips-18 can you possibly cross check this with existing jquery-match-height samples in pf-core/angular-pf and ensure we have the correct divs sized appropriately with equalizer?

priley86 commented on src/components/Cards/Card.js:
```diff
@@ -0,0 +1,59 @@
+  cardRef,
+  ...props
+}) => {
+  const equalizedCSS = {
```
again I don't know if we want to do this... I think PF Core can create this CSS eventually. I think my hope short term was that we'd be able to use Equalizer the same way we do currently with jquery-match-height. We will need feedback from @jgiardino and @jeff-phillips-18 though.

jeff-phillips-18 replied:
I would prefer to leave this out and let the application handle resizing cards to be the same height in a row. This should mean the card will take up the full height of its parent container. The application can deal with setting the appropriate parent heights. Leaving access to the `ref` is fine.

sharvit replied:
I don't think it is going to be that trivial.

Without applying flexbox behavior, you won't be able to set height for a card without damaging the card layout.

I made this sandbox to show this case:
https://codesandbox.io/s/jp25kx754y

Considering this behavior,
**shouldn't cards suppose to be a flexbox by default from patternfly-core?**

@priley86 in can't find it, but I remember @gilad215 share your equalizer code with me, and you basically handled it by equalizing the card-body itself.
It works, but if in one row you have one card with a footer and one without a footer, it breaks.

jeff-phillips-18 replied:
I'm OK with the styling to use flexbox until we get the correct styling from PF core. I would much rather see it in less/sass and not as inline styling. Does it need to be an option or should it always use flexbox?

sharvit replied:
Yes, I believe it should always use flexbox.
It will allow consumers to easily set a height for a card.

@gilad215 can you move the styles into the less/scss as described here:
https://github.com/patternfly/patternfly-react/blob/master/CONTRIBUTING.md#adding-styling-for-your-components

priley86 commented on src/components/Cards/CardBody.js:
```diff
@@ -0,0 +1,33 @@
+
+const CardBody = ({ equalized, children, className, ...props }) => {
+  const classes = ClassNames('card-pf-body', className);
+  const equalizedBody = {
```
same thing here...

sjd78 commented:
FWIW - the `react-equalizer` package can't detect if a div's content changes requiring resizing.  This will cause a problem if any card dynamically updates its contents OR if fonts are applied to the document AFTER the resize handler is invoked.

For `ovirt-engine-dashboard` I had to use a ResizeSensor for `css-element-queries` instead of `react-equalizer`.  See https://github.com/oVirt/ovirt-engine-dashboard/commit/9816ee02129da70a2dd9b4cb039bb4642c32d584.

priley86 commented:
seems much better @sjd78 .. thanks 👍

sharvit commented:
@sjd78 Thanks for sharing that.
I would love to see the `HeightMatching` component inside the `patternfly-react` code base :+1: 
Do you think you can create a quick pr?

sharvit commented on src/components/Cards/CardContainer.js:
```diff
@@ -0,0 +1,33 @@
+import Equalizer from 'react-equalizer';
+
+const CardContainer = ({ equalized, nodes, children, className, ...props }) => {
+  const classes = ClassNames('container-fluid container-cards-pf', className);
```
I think this component should be called `CardsGrid`.
And it should be delivered with `CardsGrid.Row` and `CardsGrid.Col` and `CardsGrid.Clearfix`.

See the current `Grid` component:
https://github.com/patternfly/patternfly-react/blob/master/src/components/Grid/index.js

gilad215 commented:
- Implemented @sharvit's suggestions
- Changed `CardContainer` to `CardGrid`
- Replaced `react-equalizer` with @sjd78 `HeightMatching` component (called `CardHeightMatching` for now until we decide if its a Card component or a global Component. Cards can now change their heights dynamically when their content is changed. `Card` and `CardBody` still require `equalizedCSS` in order to scale correctly.

Taking a closer look inside the HeightMatching story, you can see the top `CardDropDownButton` is not positioned right and i think its because of Flex CSS,I made the card accented so its easy to notice. waiting for your feedback on this.

gilad215 commented:
- Added less/scss implementation for the `CardHeightMatching` functionality:
added class `card-pf-match-height` to `Card` component, added `flex:1` to existing `card-pf-body` which affects all cards, i think this should be on all cards.
- Added Height Match KNOB to the Storybook, you can now see the `CardDropdownButton` behaviour

priley86 commented:
needs rebase @gilad215 ?

jeff-phillips-18 commented:
Are we ready to take WIP off this PR? Those TODOs can be addressed in future PRs.

priley86 commented on src/components/Cards/index.js:
```diff
@@ -0,0 +1,29 @@
+import CardFooter from './CardFooter';
+import CardLink from './CardLink';
+import CardDropdownButton from './CardDropdownButton';
+import CardHeightMatching from './CardHeightMatching';
```
can we also do a simple:
```
Card.Title = CardTitle;
Card.Body = CardBody;
Card.Heading = CardHeading;
Card.Footer = CardFooter;
Card.Link = CardLink;
Card.DropdownButton = CardDropdownButton;
Card.HeightMatching = CardHeightMatching;
```
like other subclassed components in this project ;)

This gives us really nice intellisense downstream...

priley86 commented on src/components/Cards/UtilizationTrendCard/index.js:
```diff
@@ -0,0 +1,15 @@
+import UtilizationCardDetailsDesc from './UtilizationCardDetailsDesc';
+import UtilizationCardDetailsLine1 from './UtilizationCardDetailsLine1';
+import UtilizationCardDetailsLine2 from './UtilizationCardDetailsLine2';
+
```
same thing here:
```
UtilizationCard.Details = UtilizationCardDetails;
UtilizationCard.DetailsCount = UtilizationCardDetailsCount;
...
```

priley86 commented:
I am really happy w/ this as an initial implementation. There are a few other javascripty things which should get resolved if you rebase the branch and run `npm run test` with new lint rules in place (but should be fairly minor).

Not sure if we want to do another pass on the css being introduced or if this is OK for round 1.

After this...happy to see it merge. great job @gilad215 !  🖖

priley86 commented:
Oh one more thing - please change the commit to “feat(cards): adds base card, height match card, ...” 

That will ensure it gets released!

gilad215 commented:
- rebased and passed the new linter rules + created a new commit info @priley86

some VerticalNav files were changed by linter/prettify and i'm not sure why, other than that i think we are good :)

priley86 commented:
looks great to me @gilad215 - thanks for turning this so fast. Happy camper here but will wait for others to weigh in on merging this.

jeff-phillips-18 commented on less/<CardMatchHeight>.less:
```diff
@@ -0,0 +1,14 @@
+.card-pf-match-height {
```
Is there a reason to have the angle brackets in the filename?

gilad215 replied:
<img width="194" alt="screenshot 2018-02-26 16 21 19" src="https://user-images.githubusercontent.com/24938324/36675281-2e892d64-1b11-11e8-94c6-3e957f7896d1.png">

I was confused by this..fixing this shortly

jeff-phillips-18 commented:
@Rohoover @jgiardino @serenamarie125 Can we get a UX Review please?

sharvit commented on sass/patternfly-react/_CardMatchHeight.scss:
```diff
@@ -0,0 +1,17 @@
+.card-pf-match-height {
```
Shouldn't it be added to all cards?

gilad215 replied:
I think we should separate the two types of cards, giving consumers more freedom with the HeightMatching Grid (e.g. 2 match-height cards and 1 fixed height card)

gilad215 commented:
Hi @Rohoover , I compared both PF and PF-React HTML,CSS and they are identical, then i noticed there is a blank space after the 4 🥇 
<img width="72" alt="screenshot 2018-03-01 10 02 10" src="https://user-images.githubusercontent.com/24938324/36833484-2dee6c88-1d38-11e8-919c-b48b70be276b.png">

I Added it to my storybook

jgiardino commented:
For a card that has no data, I haven't seen an example in patternfly that includes the EmptyState component inside a Card component. The angular example that @serenamarie125 mentioned above of a [card with no data](https://www.patternfly.org/angular-patternfly/#/api/patternfly.charts.directive:pfUtilizationTrendChart) is probably the pattern we should follow for this case. I think we would just duplicate the css that they have included as well, since this isn't in patternfly core.

jeff-phillips-18 commented:
The necessary overrides for blank-slate-pf in a cards-pf container:

```
.cards-pf .blank-slate-pf {
  border: none;

  .blank-slate-pf-icon {
    color: $color-pf-black-300;
  }
}
```

jgiardino commented on less/CardMatchHeight.less:
```diff
@@ -0,0 +1,14 @@
+}
+.card-pf-body {
+  flex: 1;
+}
```
I suggest replacing all of this with the following instead:
```
.card-pf-match-height[style*="height"] {
  position: relative;

  .card-pf-footer {
    position: absolute;
    bottom: 0;
    width: 100%;
  }
}
```
AND you can remove the inline `position: relative;` from the `card-pf` element since it's in the css. 

The issue with setting `display: flex;` on the card is that [margins do not collapse for flex items](https://www.w3.org/TR/css-flexbox-1/#flex-containers) and the current css/html of cards are dependent on collapsing margins. The simplest work-around I could come up with is to absolute position the footer to the bottom, once the height has been defined for the card.

Also, on slack you had asked about being able to set a custom height, but I haven't seen that anywhere in patternfly, and in playing around with setting the height, it doesn't look like the current css accommodates that. For example, if the height of the card is less than the height of the body, then the body contents will overflow the card.

gilad215 replied:
Thank you for the detailed answer! ill work with all of this on Sunday! until then, do we even want to allow a consumer to create a card with custom height?

jgiardino replied:
Since the pattern doesn't support that use case today, then I don't think we should include that as an example. I don't think we need to prevent consumers from setting the height either if that's something they need. I just expect that it's on the consumer to fully test and handle any css issues that come with doing that.

jgiardino commented:
Thanks for sharing the blank slate css, @jeff-phillips-18! I tweaked it slightly  to avoid the example that I pasted below where the padding is jumbo-normous. (this is the scss version) 

```
.cards-pf .blank-slate-pf {
  border: none;
  padding: $jumbotron-padding;

  .blank-slate-pf-icon {
    color: $color-pf-black-300;
  }
}
```

![image](https://user-images.githubusercontent.com/21063328/36918710-be1c9320-1e28-11e8-962c-2afaf171288b.png)

jgiardino commented:
Thanks for making the updates to the css, @gilad215! One more suggestion for those updates is to rename the css files to be `_card.scss` and `card.less` since these files are not specific to the match height feature. 

Also, your example of the empty state includes a background color, whereas @jeff-phillips-18's example does not. Jeff, are there any other styles that we're missing here?

As for the UX of the empty state pattern, I'd like to get feedback from @serenamarie125, @Rohoover and @maryclarke... 

The example below is from @jeff-phillips-18 of a card showing the empty state pattern. These examples include an action that adds content to the cards.
![image](https://user-images.githubusercontent.com/21063328/36978711-3c27d97e-2053-11e8-8910-9528d6baf0d6.png)

The [angular example of a utilization trend card](https://www.patternfly.org/angular-patternfly/#/api/patternfly.charts.directive:pfUtilizationTrendChart) shows a different pattern where no data is present. 
![image](https://user-images.githubusercontent.com/21063328/36978931-e527aef0-2053-11e8-986f-1d52141d3749.png)

I'm assuming both patterns are valid, but just want to double-check. Is the first one only for cases where the user can add contents, and the second only for cases where they can't? If so, then I think the Empty Card example for this storybook (shown below) should be updated to include an action for adding contents. Are there any other updates that you think are needed for this pattern?

![image](https://user-images.githubusercontent.com/21063328/36979070-58d61576-2054-11e8-8ca8-0386c3d1937d.png)

maryclarke commented:
I agree @jgiardino . These seem like two valid use cases and if the plus icon is being used, there should be an action button included.

gilad215 commented:
Thanks for the feedback @jgiardino @maryclarke 
For comparison I created the two types of Empty Cards you mentioned side by side on the Storybook so it can help us decide.

- Updated css file names
- Removed first Empty Card background, added a button
- Created a new Empty Card w/Info Icon

jgiardino commented:
Thanks for adding the two examples! I'm on the fence about this one. I'm kinda leaning toward removing the "no data available" example shown below, since we don't have a concrete example to follow in patternfly. 

![image](https://user-images.githubusercontent.com/21063328/36999769-2f100ca2-208f-11e8-99aa-8189544cde7f.png)

But I'm not opposed to keeping it if there is a definite need for this now. I was chatting with @Rohoover about this example earlier, and she was suggesting that the pattern would work if we changed this text from a paragraph:
```
<p class="blank-slate-pf-info">No data available</p>
```
to a header:
```
<h4 class="h1 blank-slate-pf-title">No Data Available</h4>
```
so that it looks like this:
![image](https://user-images.githubusercontent.com/21063328/36999994-e207a36a-208f-11e8-80f1-d896dba504ae.png)

But still, my preference would be to have something in core that we can consume for this pattern.

@serenamarie125 or @maryclarke Do you have any suggestions on this? Should we go with @Rohoover's suggestion for now, and add an issue to patternfly core to get an example added which we can revisit here later?

Rohoover commented:
I would definitely like to see an empty state pattern that is flexible among different uses and represented in core.

jgiardino commented:
After chatting with other team members, we'd like to move forward with @Rohoover's suggestion for the No Data Available state (see the notes in the previous comment for moving the text from a `<p>` to a `<h4>`).

We will track the design for this in the core repo and open another PR here later if needed.

gilad215 commented:
- Added [Utilization Bar Card](https://rawgit.com/gilad215/patternfly-react/storybook/feature/cards/index.html?selectedKind=Cards&selectedStory=Utilization%20Bar%20Card&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
- Fixed Documentation link bug

priley86 commented:
I would still like to try to consume this in V2V if possible. Thanks for all the hard work on this @gilad215 !!!

jeff-phillips-18 commented:
@patternfly/patternfly-react-ux

jeff-phillips-18 commented:
@gilad215 sorry didn't notice this before but I think we should try to avoid the lodash dependency. I believe @priley86 has an alternative to the lodash debounce function.

priley86 commented:
^^ great catch @jeff-phillips-18 ... i totally missed this...

@gilad215 would you mind introducing some kind of pure es6 helper for the lodash debounce function (and removing that lodash). Maybe in `src/common/helpers.js` some kind of new es6 debouncer:
```
const debounce = (fn, time) => {
  let timeout;

  return function() {
    const functionCall = () => fn.apply(this, arguments);
    
    clearTimeout(timeout);
    timeout = setTimeout(functionCall, time);
  }
}
```
There is a great article on this...
https://medium.com/@TCAS3/debounce-deep-dive-javascript-es6-e6f8d983b7a1

gilad215 commented:
- removed lodash package, created a `lodash/debounce` function inside `src/common/helpers.js`

priley86 commented:
pulled locally just now... 
tests pass 👍 
card match height debouncer working 👍

priley86 commented:
woohoo! 🎉 🎉
